pki/blacklist/blacklist.go

```diff
@@ -0,0 +1,241 @@
+)
+
+type Blacklist interface {
+	LastUpdated() time.Time
+	Update() error
+	URL() string
+	ValidateCert(cert *x509.Certificate) error
+}
+
+// blacklistImpl implements arbitrary certificate rejection using issuer and serial number tuples
+type blacklistImpl struct {
+	// url specifies the URL where the blacklist is downloaded
+	url string
+
+	// entries is the decoded entries from the downloaded blacklist
+	entries atomic.Pointer[[]blacklistEntry]
+
+	// trustedKey is an Ed25519 key which must sign the blacklist
+	trustedKey jwk.Key
+
+	// lastUpdated contains the time the certificate was last updated
+	lastUpdated time.Time
+}
+
+// blacklistEntry contains parameters for an X.509 certificate that must not be accepted for TLS connections
+type blacklistEntry struct {
+	// Issuer is a string representation (x509.Certificate.Issuer.String()) of the certificate
+	Issuer string
+
+	// SerialNumber is a string representation (x509.Certificate.SerialNumber.String()) of the certificate
+	SerialNumber string
+
+	// JWKThumbprint is an identifier of the public key per https://www.rfc-editor.org/rfc/rfc7638
+	JWKThumbprint string
+}
+
+// New creates a blacklist with the specified configuration
+func New(config config.Config) (Blacklist, error) {
+	// "Disable" (operate in a NOP mode) the blacklist when the URL is empty
+	if config.URL == "" {
+		// Return the new blacklist and a nil error
+		return &blacklistImpl{
+			trustedKey: nil,
+			url:        "",
+		}, nil
+	}
+
+	// Parse the trusted key
+	key, err := jwk.ParseKey([]byte(config.TrustedSigner), jwk.WithPEM(true))
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse key: %w", err)
+	}
+
+	// Return the new blacklist and a nil error
+	return &blacklistImpl{
+		trustedKey: key,
+		url:        config.URL,
+	}, nil
+}
+
+// ValidateCert checks for the issuer and serialNumber in the blacklist, returning nil when the cert is permitted
+// or an error when the cert is blacklisted or the blacklist cannot be retrieved
+func (b *blacklistImpl) ValidateCert(cert *x509.Certificate) error {
+	// Blacklists with an empty URL are a NOP
+	if b.URL() == "" {
+		return nil
+	}
+
+	// Extract the necessary information from the certificate
+	issuer := cert.Issuer.String()
+	serialNumber := cert.SerialNumber.String()
+
+	// If the blacklist has not yet been downloaded, do so now
+	if b.lastUpdated.IsZero() {
+		// Trigger an update of the blacklist
+		if err := b.Update(); err != nil {
+			// If the blacklist download failed then log a message about it
+			logger().WithError(err).
+				WithField("Issuer", issuer).
+				WithField("S/N", serialNumber).
+				Error("cert validation failed because the blacklist cannot be downloaded")
+
+			// Return an error indicating the blacklist cannot be retrieved
+			return ErrBlacklistMissing
+		}
+	}
+
+	// Fetch the entries
+	entriesPtr := b.entries.Load()
+
+	// The entries pointer should not be empty because of the lastUpdated check above
+	if entriesPtr == nil {
```

Comment:
Program error in nuts-node, so panic with nil deref is probably better (otherwise we'll have a very hard time finding the cause)

Reply:
I added a log message